### PR TITLE
stb_ds: fix shputs() for strdup and arena hash table

### PR DESCRIPTION
**Pull request by srdjanstipic**
This commit fixes the issue of inserting a struct to strdup and arena hastable.
Downside of this change is that it introduces additional instructions (also for default string hashtable)
but inserted code should have negligible overhead because it is easy to predict.

Feel free to reject this change if you think that it does not make sense
or it is not worth additional complexity.

### CODE REVIEW

**Comment by nothings** [comment]
Yeah, I was thinking about whether this would really be that hacky last night, and given the existing temp, it's not so bad. The only thing I would change is to just always write the key regardless, rather than bothering to test the hash table type.

**Comment by srdjanstipic** [comment]
My initial solution was always writing the key but I thought that we could minimize number of writes.
I will modify the PR to update the key always.

File: stb_ds.h [hunk 1]
```diff
@@ -587,7 +588,8 @@ extern void * stbds_shmode_func(size_t elemsize, int mode);
 
 #define stbds_shputs(t, s) \
     ((t) = stbds_hmput_key_wrapper((t), sizeof *(t), (void*) (s).key, sizeof (s).key, STBDS_HM_STRING), \
-     (t)[stbds_temp((t)-1)] = (s))
+     (t)[stbds_temp((t)-1)] = (s), \
+     (t)[stbds_temp((t)-1)].key = stbds_temp_key((t)-1))
```
Comment (srdjanstipic) [inline stb_ds.h:592]: Update the key.

File: stb_ds.h [hunk 2]
```diff
@@ -1410,9 +1413,9 @@ void *stbds_hmput_key(void *a, size_t elemsize, void *key, size_t keysize, int m
       stbds_temp(a) = i-1;
 
       switch (table->string.mode) {
-         case STBDS_SH_STRDUP:  *(char **) ((char *) a + elemsize*i) = stbds_strdup((char*) key); break;
-         case STBDS_SH_ARENA:   *(char **) ((char *) a + elemsize*i) = stbds_stralloc(&table->string, (char*)key); break;
-         case STBDS_SH_DEFAULT: *(char **) ((char *) a + elemsize*i) = (char *) key; break;
+         case STBDS_SH_STRDUP:  stbds_temp_key(a) = *(char **) ((char *) a + elemsize*i) = stbds_strdup((char*) key); break;
+         case STBDS_SH_ARENA:   stbds_temp_key(a) = *(char **) ((char *) a + elemsize*i) = stbds_stralloc(&table->string, (char*)key); break;
+         case STBDS_SH_DEFAULT: stbds_temp_key(a) = *(char **) ((char *) a + elemsize*i) = (char *) key; break;
```
Comment (srdjanstipic) [inline stb_ds.h:1418]: Save the key in all sh cases.

**Comment by srdjanstipic** [comment]
> The only thing I would change is to just always write the key regardless,
> rather than bothering to test the hash table type.

Done

**Comment by nothings** [comment]
The cost of writing the key is small since we just wrote the structure, so it will still be in cache, and loading the temp key is cheap since we just wrote it into the temp field during the function call. Of course the cost the other way is also small, as the branch will always predict correctly. So the decisive factor is that it's slightly less code to just always do it.

**Comment by srdjanstipic** [comment]
> So the decisive factor is that it's slightly less code to just always do it

I agree.

Should I fix some other things in the PR?

**Comment by nothings** [comment]
no, it's fine. I only use github for releasing, so you won't see the PR merged until the next release.

**Comment by nothings** [comment]
I moved the temp_key storage into the hash_table structure because the existing struct is used by arrays as well as hash tables, and I didn't want arrays to have to pay the storage overhead. This makes loading the temp key require one more indirection, but it should still all be in cache so it still doesn't matter.

**Pull request merged**